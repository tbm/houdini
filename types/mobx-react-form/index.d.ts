// License: LGPL-3.0-or-later

interface FieldHooks{
    sync?(e:Field): any
    onChange?(e:Field): any
    onToggle?(e:Field): any
    onFocus?(e:Field): any
    onBlur?(e:Field): any
    onSubmit?(e:Field): any
    onClear?(e:Field): any
    onReset?(e:Field): any
    onAdd?(e:Field): any
    onDel?(e:Field): any
    onSuccess?(e:Field):any
    onError?(e:Field):any

}

export declare class Field implements  SharedFieldFormMethods, FieldProperties, FieldMethods, FieldHandlers {
    readonly bindings: string;
    readonly changed: boolean;
    readonly default: boolean;
    readonly disabled: boolean;
    readonly error: string;
    readonly extra: any;
    readonly files: string;
    readonly focused: boolean;
    readonly hasError: string & boolean;
    readonly hasIncrementalKeys: boolean;
    readonly hasNestedFields: boolean;
    readonly interceptors: object;
    readonly isDefault: boolean;
    readonly isDirty: boolean;
    readonly isEmpty: boolean;
    readonly isPristine: boolean;
    readonly isValid: boolean;
    readonly id: string
    readonly key: string;
    readonly label: string;
    readonly name: string;
    readonly observers: object;
    readonly options: any;
    readonly path: string;
    readonly placeholder: string;
    readonly related: any;
    readonly rules: any;
    readonly size: number;
    readonly submitting: boolean;
    readonly touched: boolean;
    readonly type: string;
    readonly validateWith: string;
    readonly validating: boolean;
    readonly validators: any;
    readonly value: any;

    readonly state: any;

    $(fieldName: string): Field;

    add(obj: any): any;

    bind(): object;

    check(computed: string, deep?: boolean): boolean;

    clear();

    container(): Form

    del(key: any);

    each(callback: (i: Field) => void);

    focus();

    get();
    get(prop: any);

    has(key: string): boolean;

    hasIncrementalKeys(): boolean;

    hasNestedFields(): boolean;

    intercept(obj: any);

    invalidate(msg: string);
    invalidate(msg?: string);

    map(callback: (i: Field) => void);

    observe(obj: any);

    reset();

    resetValidation();

    select(path: string): Field;

    set(val: any);
    set(prop: string, val: any);

    showErrors(bool: boolean);

    submit(obj?: any);

    update(obj: any): void;

    validate(): Promise<Field>;

    onAdd(e: Field): any;

    onChange(e: Field): any;

    onClear(e: Field): any;

    onDel(e: Field): any;

    onError(e: Field): any;

    onReset(e: Field): any;

    onSubmit(e: any): any;

    onSuccess(e: Field): any;

    onToggle(e: Field): any;

    sync(e: Field): any;

    
   
   
}

interface FormHooks {
    onSubmit?(e:Field): any
    onClear?(e:Field): any
    onReset?(e:Field): any
    onAdd?(e:Field): any
    onDel?(e:Field): any
    onSuccess?(e:Field):any
    onError?(e:Field):any
}

interface FieldHandlers {
    sync?(e:Field):any
    onChange(e:Field):any
    onToggle(e:Field):any
    onSubmit?(e:any): any
    onClear?(e:Field): any
    onReset?(e:Field): any
    onAdd?(e:Field): any
    onDel?(e:Field): any
    onSuccess?(e:Field):any
    onError?(e:Field):any
}

interface FieldDefinition {
    name: string
    label?: string
    value?: any
    default?: any
    placeholder?: any
    disabled?: boolean
    related?: any
    bindings?: any
    type?: string
    options?: any
    extra?: any
    hooks?: FieldHooks
    handlers?: FieldHandlers
    fields?: Array<FieldDefinition>
    rules?: string
    id?:string
}

interface FieldProperties{
    readonly submitting: 	boolean
    readonly validating: 	boolean
    readonly bindings: 	string
    readonly observers: 	object
    readonly interceptors: 	object
    readonly size: number
    readonly path: 	string
    readonly key: 	string
    readonly name: 	string
    readonly type:	string
    readonly label: 	string
    readonly placeholder: 	string
    readonly default: 	boolean	
    readonly disabled: 	boolean
    readonly value: 	any
    readonly focused: 	boolean
    readonly touched: 	boolean
    readonly changed: 	boolean
    readonly related: any
    readonly rules: 	any
    readonly validators: any
    readonly validateWith: 	string
    readonly isValid: 	boolean
    readonly isDirty: 	boolean
    readonly isPristine: 	boolean
    readonly isDefault:	boolean
    readonly isEmpty: 	boolean
    readonly hasError: 	boolean
    readonly error: 	string
    readonly options: 	any
    readonly extra: 	any
    readonly files: 	string
    readonly hasNestedFields: 	boolean
    readonly hasIncrementalKeys: 	boolean
}

interface FieldMethods {
    bind():object
    clear()
    reset()
    focus()
    validate() : Promise<Field>
    invalidate(msg?:string)
    resetValidation()
    showErrors(bool:boolean)
}

interface FieldHooks {
    
}



interface FormInitializer{
    fields?: Array<FieldDefinition>
}

interface initializationDefinition {
    fields?:FieldDefinitions[]
}

export class Form implements SharedFieldFormMethods {

    
    constructor(definition:initializationDefinition, options?:any)
    plugins(): void
    setup(): any
    onInit(): void
    onSuccess(form:Form)
    onError(form:Form)
    onChange(values:any)
    onSubmit(...args)

    $(fieldName: string): Field;

    add(obj: any): any;

    check(computed: string, deep?: boolean): boolean;

    del(key: any);

    each(callback: (i: Field) => void);

    get();
    get(prop: any);

    has(key: string): boolean;

    hasIncrementalKeys(): boolean;

    hasNestedFields(): boolean;

    intercept(obj: any);

    invalidate(msg: string);

    map(callback: (i: Field) => void);

    observe(obj: any);

    select(path: string): Field;

    set(val: any);
    set(prop: string, val: any);

    submit(obj?: any);

    update(obj: any): void;

    readonly submitting: boolean;
    
}



interface SharedFieldFormMethods {
    update(obj:any): void
    select(path:string): 	Field
    submit(obj:any)
    check(computed:string,deep?:boolean): boolean
    get()
    get(prop:any)
    set(val:any)
    set(obj:any)
    set(prop:string, val:any) 
    has(key:string):boolean
    map(callback:(i:Field) => void)
    each(callback:(i:Field) => void)
    add(obj:any):any
    del(key:any)
    observe(obj:any)
    intercept(obj:any)
    hasNestedFields(): boolean
    hasIncrementalKeys() : boolean
    invalidate(msg:string)
    $(fieldName:string) : Field
}