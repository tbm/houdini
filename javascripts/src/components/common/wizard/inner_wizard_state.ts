// License: LGPL-3.0-or-later
import { observable, action, computed, runInAction } from "mobx";
import _ = require("lodash");

export interface TabInfoProps {
    readonly id: string
    
    /**
     * A identifier for the this tab.in this wizard. Only has to be unique
     * within this wizard instance
     * @type string
     * @memberof TabInfo
     */
    readonly tabName: string

    readonly label:string
    
    /**
     * Whether the tab is enabled for selection or not. Used mostly
     * for preventing moving to a new tab except when the previous ones are
     * validated
     * @type boolean
     * @memberof TabInfo
     */
    readonly enabled:boolean

    /**
     * Whether this tab is active, i.e. shown.
     * @type boolean
     * @memberof TabInfo
     */
    readonly active:boolean
}

export class TabInfo implements TabInfoProps{
    constructor(id:string,tabName:string,  label:string, enabled:boolean=false, active:boolean=false){
        runInAction(() =>{
        this.id = id
        this.tabName = tabName
        this.enabled = enabled
        this.active = active
        this.label = label
    })
    }
    /**
     * Represents the document-unique id of the tab.
     * @type string
     * @memberof TabInfo
     */
    @observable id: string
    
    /**
     * A identifier for the this tab.in this wizard. Only has to be unique
     * within this wizard instance
     * @type string
     * @memberof TabInfo
     */
    @observable tabName: string

    @observable label:string
    
    /**
     * Whether the tab is enabled for selection or not. Used mostly
     * for preventing moving to a new tab except when the previous ones are
     * validated
     * @type boolean
     * @memberof TabInfo
     */
    @observable enabled:boolean

    /**
     * Whether this tab is active, i.e. shown.
     * @type boolean
     * @memberof TabInfo
     */
    @observable active:boolean
    
    
}

/**
 * A MobX managed state for handling all the state of the Wizard
 * @export
 * @class WizardState
 */
export class InnerWizardState {

    @observable tabs : Array<TabInfo> = new Array<TabInfo>()

    
    /**
     * Add a new tab. If it's the first, mark it as active and enabled.
     * @param  {string} tabName the name of the tab to add
     * @return {TabInfoProps} the tab added
     * @memberof WizardState
     */
    @action.bound
    addTab(tabName:string, label:string) : TabInfoProps {
        var self = this
        var newTab = new TabInfo(_.uniqueId('tab'), tabName, label)
        if (self.tabs.length == 0){
            newTab.enabled = true
            newTab.active = true
        }
        
        self.tabs.push(newTab)
        return newTab
    }

    /**
     * Set the enabled status of a tab. If a tab is disabled, all of the tabs
     * after it are as well. If a tab is enabled, all of the tabs before it are
     * as well.
     * @param  {string|TabInfoProps} tab the tabname or reference to the tab 
     * you want to change the validation state for
     * @param  {boolean} enabled Whether to enabled or disable the tab
     * @return {void}
     * @memberof WizardState
     */
    @action.bound
    updateValidationState(tab:string|TabInfoProps, enabled: boolean)
    {
        let self = this
        let indexOfOurState = -1
        if (typeof tab === "string")
        {
            indexOfOurState = _.findIndex(self.tabs, (o) => o.tabName == tab)
        }
        else {
            indexOfOurState = _.findIndex(self.tabs, (o) => o == tab)
        }
        if (indexOfOurState != -1) {
            if (enabled){
                _.slice(self.tabs, 0, indexOfOurState+1).forEach((o) => o.enabled = true)
                _.slice(self.tabs, indexOfOurState+1).forEach((o) => o.enabled = false)
            }
            else
            {
                _.slice(self.tabs, 0, indexOfOurState).forEach((o) => o.enabled = true)
                _.slice(self.tabs, indexOfOurState).forEach((o) => o.enabled = false)
            }

        }
    }

    @action.bound
    getTabByName(tabName:string):TabInfoProps{
        return _.find(this.tabs, (i) => i.tabName === tabName)
    }
    
    @action.bound
    private getTabInfo(tabId:string|TabInfoProps): TabInfo {
        if (typeof tabId === 'string'){
            return _.find(this.tabs, (i) => i.id === tabId)
        }
        else{
            return _.find(this.tabs, (i) => i == tabId)
        }

        
    }

    @action.bound
    peekNextTab(tab?:TabInfo):TabInfo{
        if (!tab)
        {
            tab = this.activeTab
        }
        if (tab)
        {
            let i = _.findIndex(this.tabs, tab)
            if (i < 0)
            {
                return null
            }
            let nextTabIndex =i +1
            if ( nextTabIndex >= this.tabs.length)
            {
                return null
            }
            return this.tabs[nextTabIndex]
        }

        return null
    }

    /**
     * Set a new tab as active. If the tab to be made active is disabled, 
     * we do nothing.
     * @param  {string|TabInfo} tabNameId the tab to set as active
     * @return {TabInfoProps} the tab you made active or null if the tab wasn't
     * found or is disabled
     * @memberof WizardState
     */
    @action.bound
    handleTabChange(tabId:string|TabInfoProps):TabInfoProps {
        let self = this
        
        let tabInfo = self.getTabInfo(tabId)
        if (tabInfo && tabInfo.enabled) {
            _.forEach(_.without(self.tabs, tabInfo), (o) => runInAction(() => o.active = false))
            tabInfo.active = true
            return tabInfo
        }
        return null
    }

    // /**
    //  * Set the next tab as active. If the next tab is disabled, does nothing.
    //  * @return {TabInfoProps} that tab you've now moved to or null if none exists
    //  * @memberof WizardState
    //  */
    // @action
    // nextTab() : TabInfoProps
    // {
    //
    // }

    /**
     * Get the currently active tab
     * @readonly
     * @type TabInfo tab info for the currently active tab
     * @memberof WizardState
     */
    @computed
    get activeTab():TabInfoProps {

        return _.find(this.tabs, (o) => o.active )
    }
}