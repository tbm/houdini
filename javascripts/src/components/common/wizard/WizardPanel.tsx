// License: LGPL-3.0-or-later
import * as React from 'react';
import { TabPanel } from "react-aria-tabpanel";
import { InnerWizardState } from "./inner_wizard_state";
import {  observer, Provider } from 'mobx-react'
import { Form, Field } from 'mobx-react-form';
import {WizardState} from './wizard_state';
import {computed} from 'mobx';

export interface WizardPanelProps {
    tabName: string

    wizardState: WizardState
    form:Field

    [props:string]:any
}

@observer
export class WizardPanel extends React.Component<WizardPanelProps, {}> {
    @computed
    get tab(){
        return this.props.wizardState.tabsByName[this.props.tabName]
    }
    @computed
    get isActive(){
        return this.tab.active
    }
    render() {

        let props = this.props.props ? this.props.props : {}
        let style = this.isActive ? {display: "block"} : {display: "none"}
        return <TabPanel tabId={this.tab.id} active={this.isActive}
        {...props} className="wizard-step" style={style}>
                <fieldset>

                        {this.props.children}

                </fieldset>
        </TabPanel>
    }
}

