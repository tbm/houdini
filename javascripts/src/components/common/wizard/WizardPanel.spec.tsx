// License: LGPL-3.0-or-later
import * as React from 'react';
import 'jest';

import * as Component from './WizardPanel'
import { InnerWizardState } from './inner_wizard_state';
import {Form, Field} from 'mobx-react-form';
import { shallow, render } from 'enzyme';
import {TabPanel, Wrapper} from 'react-aria-tabpanel'
import toJson from 'enzyme-to-json';
const TestRenderer = require('react-test-renderer')

describe('WizardPanel', () => {
    test('shallow render', () => {
        let fields = [{name: 'fun', id: 'fun', label: 'alsofun'}]
        const form = new Form({fields});
        
        const ws = new InnerWizardState()
        
        ws.addTab('something', 'something lable')
        const tree = shallow(<Component.WizardPanel tabId="something"
        wizardState={ws} form={form} subForm={form.$('fun')}/>)
        
        
        expect(toJson(tree)).toMatchSnapshot()
    })

    test('deep render', () =>{
        let fields = [{name: 'fun', id: 'fun', label: 'alsofun'}]
        const form = new Form({fields});
        
        const ws = new InnerWizardState()
        
        ws.addTab('something', 'something lable')
        const tree = TestRenderer.create(<Wrapper onChange={ws.handleTabChange}><Component.WizardPanel tabId="something"
        wizardState={ws} form={form} subForm={form.$('fun')}/></Wrapper>)
        
        
        expect(tree.toJSON()).toMatchSnapshot()
    })
})