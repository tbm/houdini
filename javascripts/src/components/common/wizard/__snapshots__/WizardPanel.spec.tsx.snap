// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WizardPanel deep render 1`] = `
<div>
  <div
    aria-describedby="something"
    aria-hidden={true}
    className="wizard-step"
    id="something-panel"
    onKeyDown={[Function]}
    role="tabpanel"
    style="display: none"
  >
    <fieldset />
  </div>
</div>
`;

exports[`WizardPanel shallow render 1`] = `
<WizardPanel
  form={
    Form {
      "$handlers": Object {},
      "$hooks": Object {},
      "$submitting": false,
      "$validating": true,
      "debouncedValidation": [Function],
      "execHandler": [Function],
      "execHook": [Function],
      "fields": Object {
        "fun": Field {
          "$bindings": "default",
          "$changed": false,
          "$clearing": false,
          "$default": "",
          "$disabled": false,
          "$extra": null,
          "$focused": false,
          "$handlers": Object {},
          "$hooks": Object {},
          "$initial": "",
          "$input": [Function],
          "$interceptors": null,
          "$label": "alsofun",
          "$observers": null,
          "$options": Object {},
          "$output": [Function],
          "$placeholder": "",
          "$related": Array [],
          "$resetting": false,
          "$rules": null,
          "$submitting": false,
          "$touched": false,
          "$validatedWith": "value",
          "$validating": false,
          "$validators": null,
          "$value": "",
          "autoFocus": false,
          "debouncedValidation": [Function],
          "disposeValidationOnBlur": [Function],
          "errorAsync": null,
          "errorSync": null,
          "execHandler": [Function],
          "execHook": [Function],
          "fields": Object {},
          "files": undefined,
          "hasInitialNestedFields": false,
          "id": "fun--1",
          "incremental": false,
          "intercept": [Function],
          "key": "fun",
          "name": "fun",
          "noop": [Function],
          "observe": [Function],
          "onAdd": [Function],
          "onBlur": [Function],
          "onChange": [Function],
          "onClear": [Function],
          "onDel": [Function],
          "onDrop": [Function],
          "onFocus": [Function],
          "onReset": [Function],
          "onSubmit": [Function],
          "onToggle": [Function],
          "path": "fun",
          "showError": false,
          "state": State {
            "$struct": Array [],
            "bindings": Bindings {
              "rewriters": Object {
                "default": Object {
                  "autoFocus": "autoFocus",
                  "checked": "checked",
                  "disabled": "disabled",
                  "id": "id",
                  "label": "label",
                  "name": "name",
                  "onBlur": "onBlur",
                  "onChange": "onChange",
                  "onFocus": "onFocus",
                  "placeholder": "placeholder",
                  "type": "type",
                  "value": "value",
                },
              },
              "templates": Object {},
            },
            "current": Object {
              "fields": Object {},
              "props": Object {},
            },
            "disposers": Object {
              "interceptor": Object {},
              "observer": Object {},
            },
            "form": [Circular],
            "initial": Object {
              "fields": Object {},
              "props": Object {},
            },
            "mode": "unified",
            "options": Options {
              "options": Object {
                "allowRequired": false,
                "autoParseNumbers": false,
                "defaultGenericError": null,
                "retrieveOnlyDirtyValues": false,
                "retrieveOnlyEnabledFields": false,
                "showErrorsOnBlur": true,
                "showErrorsOnChange": true,
                "showErrorsOnClear": false,
                "showErrorsOnInit": false,
                "showErrorsOnReset": true,
                "showErrorsOnSubmit": true,
                "strictDelete": true,
                "strictUpdate": false,
                "submitThrowsError": true,
                "uniqueId": [Function],
                "validateDisabledFields": false,
                "validateOnBlur": true,
                "validateOnChange": false,
                "validateOnInit": true,
                "validationDebounceOptions": Object {
                  "leading": false,
                  "trailing": true,
                },
                "validationDebounceWait": 250,
              },
            },
            "strict": false,
          },
          "sync": [Function],
          "type": "text",
          "validationAsyncData": Object {},
          "validationErrorStack": Array [],
          "validationFunctionsData": Array [],
        },
      },
      "intercept": [Function],
      "name": null,
      "noop": [Function],
      "observe": [Function],
      "onAdd": [Function],
      "onClear": [Function],
      "onDel": [Function],
      "onReset": [Function],
      "onSubmit": [Function],
      "state": State {
        "$struct": Array [],
        "bindings": Bindings {
          "rewriters": Object {
            "default": Object {
              "autoFocus": "autoFocus",
              "checked": "checked",
              "disabled": "disabled",
              "id": "id",
              "label": "label",
              "name": "name",
              "onBlur": "onBlur",
              "onChange": "onChange",
              "onFocus": "onFocus",
              "placeholder": "placeholder",
              "type": "type",
              "value": "value",
            },
          },
          "templates": Object {},
        },
        "current": Object {
          "fields": Object {},
          "props": Object {},
        },
        "disposers": Object {
          "interceptor": Object {},
          "observer": Object {},
        },
        "form": [Circular],
        "initial": Object {
          "fields": Object {},
          "props": Object {},
        },
        "mode": "unified",
        "options": Options {
          "options": Object {
            "allowRequired": false,
            "autoParseNumbers": false,
            "defaultGenericError": null,
            "retrieveOnlyDirtyValues": false,
            "retrieveOnlyEnabledFields": false,
            "showErrorsOnBlur": true,
            "showErrorsOnChange": true,
            "showErrorsOnClear": false,
            "showErrorsOnInit": false,
            "showErrorsOnReset": true,
            "showErrorsOnSubmit": true,
            "strictDelete": true,
            "strictUpdate": false,
            "submitThrowsError": true,
            "uniqueId": [Function],
            "validateDisabledFields": false,
            "validateOnBlur": true,
            "validateOnChange": false,
            "validateOnInit": true,
            "validationDebounceOptions": Object {
              "leading": false,
              "trailing": true,
            },
            "validationDebounceWait": 250,
          },
        },
        "strict": false,
      },
      "validator": Validator {
        "drivers": Object {
          "vjf": VJF {
            "options": Options {
              "options": Object {
                "allowRequired": false,
                "autoParseNumbers": false,
                "defaultGenericError": null,
                "retrieveOnlyDirtyValues": false,
                "retrieveOnlyEnabledFields": false,
                "showErrorsOnBlur": true,
                "showErrorsOnChange": true,
                "showErrorsOnClear": false,
                "showErrorsOnInit": false,
                "showErrorsOnReset": true,
                "showErrorsOnSubmit": true,
                "strictDelete": true,
                "strictUpdate": false,
                "submitThrowsError": true,
                "uniqueId": [Function],
                "validateDisabledFields": false,
                "validateOnBlur": true,
                "validateOnChange": false,
                "validateOnInit": true,
                "validationDebounceOptions": Object {
                  "leading": false,
                  "trailing": true,
                },
                "validationDebounceWait": 250,
              },
            },
            "promises": Array [],
            "validator": null,
          },
        },
        "error": null,
        "form": [Circular],
        "options": Object {},
        "plugins": Object {
          "dvr": false,
          "svk": false,
          "vjf": true,
        },
        "promises": Array [],
        "schema": Object {},
      },
    }
  }
  subForm={
    Field {
      "$bindings": "default",
      "$changed": false,
      "$clearing": false,
      "$default": "",
      "$disabled": false,
      "$extra": null,
      "$focused": false,
      "$handlers": Object {},
      "$hooks": Object {},
      "$initial": "",
      "$input": [Function],
      "$interceptors": null,
      "$label": "alsofun",
      "$observers": null,
      "$options": Object {},
      "$output": [Function],
      "$placeholder": "",
      "$related": Array [],
      "$resetting": false,
      "$rules": null,
      "$submitting": false,
      "$touched": false,
      "$validatedWith": "value",
      "$validating": false,
      "$validators": null,
      "$value": "",
      "autoFocus": false,
      "debouncedValidation": [Function],
      "disposeValidationOnBlur": [Function],
      "errorAsync": null,
      "errorSync": null,
      "execHandler": [Function],
      "execHook": [Function],
      "fields": Object {},
      "files": undefined,
      "hasInitialNestedFields": false,
      "id": "fun--1",
      "incremental": false,
      "intercept": [Function],
      "key": "fun",
      "name": "fun",
      "noop": [Function],
      "observe": [Function],
      "onAdd": [Function],
      "onBlur": [Function],
      "onChange": [Function],
      "onClear": [Function],
      "onDel": [Function],
      "onDrop": [Function],
      "onFocus": [Function],
      "onReset": [Function],
      "onSubmit": [Function],
      "onToggle": [Function],
      "path": "fun",
      "showError": false,
      "state": State {
        "$struct": Array [],
        "bindings": Bindings {
          "rewriters": Object {
            "default": Object {
              "autoFocus": "autoFocus",
              "checked": "checked",
              "disabled": "disabled",
              "id": "id",
              "label": "label",
              "name": "name",
              "onBlur": "onBlur",
              "onChange": "onChange",
              "onFocus": "onFocus",
              "placeholder": "placeholder",
              "type": "type",
              "value": "value",
            },
          },
          "templates": Object {},
        },
        "current": Object {
          "fields": Object {},
          "props": Object {},
        },
        "disposers": Object {
          "interceptor": Object {},
          "observer": Object {},
        },
        "form": Form {
          "$handlers": Object {},
          "$hooks": Object {},
          "$submitting": false,
          "$validating": true,
          "debouncedValidation": [Function],
          "execHandler": [Function],
          "execHook": [Function],
          "fields": Object {
            "fun": [Circular],
          },
          "intercept": [Function],
          "name": null,
          "noop": [Function],
          "observe": [Function],
          "onAdd": [Function],
          "onClear": [Function],
          "onDel": [Function],
          "onReset": [Function],
          "onSubmit": [Function],
          "state": [Circular],
          "validator": Validator {
            "drivers": Object {
              "vjf": VJF {
                "options": Options {
                  "options": Object {
                    "allowRequired": false,
                    "autoParseNumbers": false,
                    "defaultGenericError": null,
                    "retrieveOnlyDirtyValues": false,
                    "retrieveOnlyEnabledFields": false,
                    "showErrorsOnBlur": true,
                    "showErrorsOnChange": true,
                    "showErrorsOnClear": false,
                    "showErrorsOnInit": false,
                    "showErrorsOnReset": true,
                    "showErrorsOnSubmit": true,
                    "strictDelete": true,
                    "strictUpdate": false,
                    "submitThrowsError": true,
                    "uniqueId": [Function],
                    "validateDisabledFields": false,
                    "validateOnBlur": true,
                    "validateOnChange": false,
                    "validateOnInit": true,
                    "validationDebounceOptions": Object {
                      "leading": false,
                      "trailing": true,
                    },
                    "validationDebounceWait": 250,
                  },
                },
                "promises": Array [],
                "validator": null,
              },
            },
            "error": null,
            "form": [Circular],
            "options": Object {},
            "plugins": Object {
              "dvr": false,
              "svk": false,
              "vjf": true,
            },
            "promises": Array [],
            "schema": Object {},
          },
        },
        "initial": Object {
          "fields": Object {},
          "props": Object {},
        },
        "mode": "unified",
        "options": Options {
          "options": Object {
            "allowRequired": false,
            "autoParseNumbers": false,
            "defaultGenericError": null,
            "retrieveOnlyDirtyValues": false,
            "retrieveOnlyEnabledFields": false,
            "showErrorsOnBlur": true,
            "showErrorsOnChange": true,
            "showErrorsOnClear": false,
            "showErrorsOnInit": false,
            "showErrorsOnReset": true,
            "showErrorsOnSubmit": true,
            "strictDelete": true,
            "strictUpdate": false,
            "submitThrowsError": true,
            "uniqueId": [Function],
            "validateDisabledFields": false,
            "validateOnBlur": true,
            "validateOnChange": false,
            "validateOnInit": true,
            "validationDebounceOptions": Object {
              "leading": false,
              "trailing": true,
            },
            "validationDebounceWait": 250,
          },
        },
        "strict": false,
      },
      "sync": [Function],
      "type": "text",
      "validationAsyncData": Object {},
      "validationErrorStack": Array [],
      "validationFunctionsData": Array [],
    }
  }
  tabId="something"
  wizardState={
    WizardState {
      "tabs": Array [
        TabInfo {
          "active": true,
          "enabled": true,
          "id": "tab1",
          "label": "something lable",
          "tabName": "something",
        },
      ],
    }
  }
/>
`;
